### PR TITLE
Boobook recognizer etc

### PR DESCRIPTION
**Pull request by towsey**
# Title of PR
Boobook sub-branch code review
What is the purpose of this PR?

## Changes
Created new Boobook recognizer and Aust Bittern recognizers.
Also set up yml for Koalav3 recognizer.

## Issues
There appear to be a bunch of errors in Azure pipeline.
## Visual Changes
None

## Final Checklist

-   [Anthony ] Assign reviewers if you have permission
-   [ ???] Assign labels if you have permission
-   [ ???] Link issues related to PR
-   [ ???] Link any PRs or issues blocking this PR from being merged
-   [ ???] Remove/Reduce warnings from edited files
-   [ One unit test for Flying Fox fixed.] Unit tests have been added for changes
-   [ Code builds] Ensure CI build is passing

### CODE REVIEW

**Comment by towsey** [comment]
I fixed the conflict in one file.

File: src/AnalysisPrograms/Recognizers/Base/RecognizerBase.cs [hunk 1]
```diff
@@ -292,7 +298,9 @@ public override AnalysisResult2 Analyze<T>(AnalysisSettings analysisSettings, Se
 
         public override void WriteEventsFile(FileInfo destination, IEnumerable<EventBase> results)
         {
-            Csv.WriteToCsv(destination, results.Select(x => (AcousticEvent)x));
+            // TODO check this
+            //Csv.WriteToCsv(destination, results.Select(x => (AcousticEvent)x));
+            Csv.WriteToCsv(destination, results.Select(x => (EventCommon)x));
```
Comment (atruskie) [inline src/AnalysisPrograms/Recognizers/Base/RecognizerBase.cs:303]: I think we should also write out the JSON variant here

Comment (atruskie) [reply src/AnalysisPrograms/Recognizers/Base/RecognizerBase.cs:303]: No cast will be necessary though

File: src/AnalysisPrograms/Recognizers/Birds/BotaurusPoiciloptilus.cs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+            // This call sets a restriction so that only one generic algorithm is used.
+            // CHANGE this to accept multiple generic algorithms as required.
+            //if (result.Profiles.SingleOrDefault() is OnebinTrackParameters)
+            if (config.Profiles?.Count == 1 && config.Profiles.First().Value is OnebinTrackParameters)
+            {
+                return config;
+            }
+
+            throw new ConfigFileException("Autralasian Bittern expects one and only one ForwardTrack algorithm.", file);
```
Comment (atruskie) [inline src/AnalysisPrograms/Recognizers/Birds/BotaurusPoiciloptilus.cs:61]:
```suggestion
            throw new ConfigFileException($"Autralasian Bittern expects one and only one {nameof(OneBinTrackParameters)}} algorithm.", file);
```

File: src/AnalysisPrograms/Recognizers/KoalaV3.cs [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            GenericRecognizer.ValidateProfileTagsMatchAlgorithms(config.Profiles, file);
+
+            // This call sets a restriction so that only one generic algorithm is used.
+            // CHANGE this to accept multiple generic algorithms as required.
+            //if (result.Profiles.SingleOrDefault() is ForwardTrackParameters)
+            if (config.Profiles?.Count == 1 && config.Profiles.First().Value is ForwardTrackParameters)
+            {
+                return config;
+            }
+
```
Comment (atruskie) [inline src/AnalysisPrograms/Recognizers/KoalaV3.cs:57]: Is this recogniser in complete? I thought you wanted to detect harmonics as well?

File: src/AudioAnalysisTools/AudioAnalysisTools.csproj [hunk 4]
```diff
@@ -36,4 +36,7 @@
     </PackageReference>
     <PackageReference Include="Towel" Version="1.0.11-alpha" />
   </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Events\Tracks\" />
+  </ItemGroup>
```
Comment (atruskie) [inline src/AudioAnalysisTools/AudioAnalysisTools.csproj:41]: This is weird. If I'm reading this right it looks like we've added an empty folder to the solution?

Check the folder is empty, and if it is, remove these lines

File: src/AudioAnalysisTools/Events/EventCommon.cs [hunk 5]
```diff
@@ -47,6 +50,11 @@ public abstract class EventCommon : EventBase, IDrawableEvent
         /// </summary>
         public Interval<double> ScoreRange { get; set; }
 
+        /// <summary>
+        /// Gets a score in the range 0,1.
+        /// Up to user to determine a suitable range maximum.
+        /// Minimum of range assumed to be zero.
+        /// </summary>
         public double ScoreNormalised => this.Score / this.ScoreRange.Maximum;
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/EventCommon.cs:58]: The calculation of `ScoreNormalized` is only partially complete. Use the lower bound of the provided interval in the calculation and remove the assumption of zero.

File: src/AudioAnalysisTools/Events/Types/CompositeEvent.cs [hunk 6]
```diff
@@ -57,6 +57,19 @@ public IEnumerable<ITrack> Tracks
...
+            foreach (var @event in this.ComponentEvents)
+            {
+                sum += @event.Score;
+            }
+
+            var score = sum / this.ComponentEvents.Count;
+
+            return score / this.ScoreRange.Maximum;
+        }
+
```
Comment (atruskie) [inline src/AudioAnalysisTools/Events/Types/CompositeEvent.cs:72]:
```suggestion
        public override double ScoreNormalised => this.ComponentEvents.Average()
```

1. Should be override
2. Should be property
3. Why not use score normalized from component events?
4. Using Score will almost certainly calculate an incorrect result if the range maximum's are different

File: src/AudioAnalysisTools/UnitConverters.cs [hunk 7]
```diff
@@ -127,7 +127,7 @@ public UnitConverters(double segmentStartOffset, int sampleRate, int frameSize,
         /// <remarks>
         /// Measured in seconds per spectrogram frame.
         /// </remarks>
-        public LinearSecondsScale SecondsScale { get; }
+        //public LinearSecondsScale SecondsScale { get; }
```
Comment (atruskie) [inline src/AudioAnalysisTools/UnitConverters.cs:130]: remove this line, it is not used, and further is related to a flawed understanding

File: src/AnalysisPrograms/Recognizers/Base/RecognizerBase.cs (hunk 1 above)
Comment (towsey) [reply src/AnalysisPrograms/Recognizers/Base/RecognizerBase.cs:303]: Good - but I don't know how to do this.

File: src/AnalysisPrograms/Recognizers/KoalaV3.cs (hunk 3 above)
Comment (towsey) [reply src/AnalysisPrograms/Recognizers/KoalaV3.cs:57]: This is work in progress. I should have left a message to this effect at top of class.

File: src/AudioAnalysisTools/AudioAnalysisTools.csproj (hunk 4 above)
Comment (towsey) [reply src/AudioAnalysisTools/AudioAnalysisTools.csproj:41]: The folder was empty and I have removed it.

File: src/AudioAnalysisTools/UnitConverters.cs (hunk 7 above)
Comment (towsey) [reply src/AudioAnalysisTools/UnitConverters.cs:130]: THis line and comment have been removed.

File: src/AnalysisPrograms/Recognizers/Birds/BotaurusPoiciloptilus.cs (hunk 2 above)
Comment (towsey) [reply src/AnalysisPrograms/Recognizers/Birds/BotaurusPoiciloptilus.cs]: I have fixed the class name. Suggestion done.

**Comment by towsey** [comment]
I believe I have made all the requested changes.

File: src/AnalysisPrograms/Recognizers/Base/RecognizerBase.cs (hunk 1 above)
Comment (atruskie) [reply src/AnalysisPrograms/Recognizers/Base/RecognizerBase.cs:303]: I've done it

**Pull request merged**